pkg/alertmanager/multitenant.go

```diff
@@ -250,29 +290,175 @@ func createMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, fallbackC
+	if err := am.loadAndSyncConfigs(ctx, reasonInitial); err != nil {
+		return err
+	}
+
+	if am.cfg.ShardingEnabled {
+		// With the initial sync now completed, we should have loaded all assigned alertmanager configurations to this instance. We can switch it to ACTIVE and start serving requests.
+		if err := am.ringLifecycler.ChangeState(ctx, ring.ACTIVE); err != nil {
+			return errors.Wrapf(err, "switch instance to %s in the ring", ring.ACTIVE)
+		}
+
+		// Wait until the ring client detected this instance in the ACTIVE state.
+		level.Info(am.logger).Log("msg", "waiting until alertmanager is ACTIVE in the ring")
+		if err := ring.WaitInstanceState(ctx, am.ring, am.ringLifecycler.GetInstanceID(), ring.ACTIVE); err != nil {
+			return err
+		}
+		level.Info(am.logger).Log("msg", "alertmanager is ACTIVE in the ring")
+	}
+
 	return nil
 }
 
-func (am *MultitenantAlertmanager) iteration(ctx context.Context) error {
-	err := am.updateConfigs()
+func (am *MultitenantAlertmanager) run(ctx context.Context) error {
+	tick := time.NewTicker(am.cfg.PollInterval)
+	defer tick.Stop()
+
+	var ringTickerChan <-chan time.Time
+	var ringLastState ring.ReplicationSet
+
+	if am.cfg.ShardingEnabled {
+		ringLastState, _ = am.ring.GetAllHealthy(RingOp)
+		ringTicker := time.NewTicker(util.DurationWithJitter(am.cfg.ShardingRing.RingCheckPeriod, 0.2))
+		defer ringTicker.Stop()
+		ringTickerChan = ringTicker.C
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-am.subservicesWatcher.Chan():
+			return errors.Wrap(err, "alertmanager subservices failed")
+		case <-tick.C:
+			// We don't want to halt execution here but instead just log what happened.
+			if err := am.loadAndSyncConfigs(ctx, reasonPeriodic); err != nil {
+				level.Warn(am.logger).Log("msg", "error while synchronizing alertmanager configs", "err", err)
+			}
+		case <-ringTickerChan:
+			// We ignore the error because in case of error it will return an empty
+			// replication set which we use to compare with the previous state.
+			currRingState, _ := am.ring.GetAllHealthy(RingOp)
+
+			if ring.HasReplicationSetChanged(ringLastState, currRingState) {
+				ringLastState = currRingState
+				if err := am.loadAndSyncConfigs(ctx, reasonRingChange); err != nil {
+					level.Warn(am.logger).Log("msg", "error while synchronizing alertmanager configs", "err", err)
+				}
+			}
+		}
+	}
+}
+
+func (am *MultitenantAlertmanager) loadAndSyncConfigs(ctx context.Context, syncReason string) error {
+	level.Info(am.logger).Log("msg", "synchronizing alertmanager configs for users")
+	cfgs, err := am.loadAlertmanagerConfigs(ctx)
 	if err != nil {
-		level.Warn(am.logger).Log("msg", "error updating configs", "err", err)
+		am.syncFailures.WithLabelValues(syncReason).Inc()
+		return err
 	}
-	// Returning error here would stop "MultitenantAlertmanager" service completely,
-	// so we return nil to keep service running.
+
+	am.syncConfigs(cfgs)
+	am.syncTotal.WithLabelValues(syncReason).Inc()
```

Comment:
This is metric is tracking the total number of attempts, not successes. You ned to move it before calling `loadAlertmanagerConfigs()` otherwise it doesn't match with `syncFailures`.